### PR TITLE
Add spec for encoding and double-encoding for plugin properties

### PR DESCRIPTION
**Pull request by daliwei**

### CODE REVIEW

**Comment by daliwei** [comment]
WIP

**Comment by daliwei** [comment]
@pierre Please review.

File: spec/killbill_client/encoding_spec.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        },
+        {
+            'key'   => 'details',
+            'value' =>  {
+                'eventType' => 'voidEvent',
+                'transactionType' => 'void',
+                'contractType' => 'temp'
+            }.to_json
+        }
+    ].map{|hash| KVPair.from_hash(hash)}
```
Comment (pierre) [inline spec/killbill_client/encoding_spec.rb]: Could you simplify (`KVPair` isn't necessary)?

File: spec/killbill_client/encoding_spec.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    options = {
+        :params => {:controlPluginName => "killbill-example-plugin"},
+        :pluginProperty => plugin_properties
+    }
+    http_adapter = DummyForHTTPAdapter.new
+    uri = http_adapter.send(:encode_params, options)
+    (uri == expected_uri).should be_true
```
Comment (pierre) [inline spec/killbill_client/encoding_spec.rb]: RSpec has been [upgraded](https://github.com/killbill/killbill-client-ruby/pull/22) to 3.4 in master. The syntax has changed a little bit.

File: lib/killbill_client/api/net_http_adapter.rb [hunk 3]
```diff
@@ -66,8 +56,23 @@ def request(method, relative_uri, options = {})
...
             }
             pairs.flatten!
-            uri += "?#{pairs.join '&'}"
+            query_params = "?#{pairs.join '&'}"
           end
+          query_params
+        end
+
+        def request(method, relative_uri, options = {})
+
```
Comment (pierre) [inline lib/killbill_client/api/net_http_adapter.rb]: Extra blank line.

File: lib/killbill_client/api/net_http_adapter.rb [hunk 4]
```diff
@@ -35,24 +35,14 @@ def net_http
...
+        def encode_params(options = {})
           # Plugin properties are passed in the options but we want to send them as query parameters,
           # so remove with from global hash and insert them under :params
           plugin_properties = options.delete :pluginProperty
           if plugin_properties && plugin_properties.size > 0
-            options[:params][:pluginProperty] = plugin_properties.map { |p| "#{p.key}=#{p.value}" }
+            options[:params][:pluginProperty] = plugin_properties.map { |p| "#{CGI.escape p.key}=#{CGI.escape p.value}" }
           end
-
+          query_params = ''
```
Comment (pierre) [inline lib/killbill_client/api/net_http_adapter.rb]: The flow could be simplified a bit:

```
return nil unless (options[:params] && !options[:params].empty?)

pairs = options[:params].map { |key, value|
# ...
```

And below, you could have:

```
uri += encode_params(options).to_s
```

File: spec/killbill_client/encoding_spec.rb (hunk 1 above)
Comment (daliwei) [inline spec/killbill_client/encoding_spec.rb]: @pierre Add KVPair is to be able to visit key and value as called in https://github.com/killbill/killbill-client-ruby/blob/master/lib/killbill_client/api/net_http_adapter.rb#L53

Do you have any suggestions on how to simply this?

**Comment by daliwei** [comment]
@pierre Can you take a look?

File: spec/killbill_client/encoding_spec.rb (hunk 1 above)
Comment (pierre) [inline spec/killbill_client/encoding_spec.rb]: You should use `PluginPropertyAttributes` instead, for example:

* https://github.com/killbill/killbill-client-ruby/blob/05550b1eae4bd66affbbab548579ca541f91ea60/lib/killbill_client/models/payment_method.rb#L95-L99
* https://github.com/killbill/killbill-integration-tests/blob/1067b362ea5bf5a9c9b585f76fd450902d133341/killbill-integration-tests/plugins/payment/errors-handling/plugins_error_handling_base.rb#L58-L60

File: lib/killbill_client/api/net_http_adapter.rb [hunk 5]
```diff
@@ -35,6 +35,30 @@ def net_http
...
+              internal_pairs = value.map do |simple_value|
+                "#{CGI.escape key.to_s}=#{CGI.escape simple_value.to_s}"
+              end
+              internal_pairs
+            else
+              "#{CGI.escape key.to_s}=#{CGI.escape value.to_s}"
+            end
+          }
+          pairs.flatten!
+          return "?#{pairs.join '&'}"
```
Comment (pierre) [inline lib/killbill_client/api/net_http_adapter.rb]: `return` is optional here (last statement of the method).

File: spec/killbill_client/encoding_spec.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    KVPair.new.tap do |p|
+      p.key = hash['key']
+      p.value = hash['value']
+    end
+  end
+end
+
+class DummyForHTTPAdapter
+  include KillBillClient::API::Net::HTTPAdapter
+end
```
Comment (pierre) [inline spec/killbill_client/encoding_spec.rb]: Nit: missing trailing new line.

File: spec/killbill_client/encoding_spec.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    options = {
+        :params => {:controlPluginName => "killbill-example-plugin"},
+        :pluginProperty => plugin_properties
+    }
+    http_adapter = DummyForHTTPAdapter.new
+    uri = http_adapter.send(:encode_params, options)
+    expect(uri).to eq(expected_uri)
+  end
+
+  def expected_uri
```
Comment (pierre) [inline spec/killbill_client/encoding_spec.rb]: Consider using [let](https://github.com/killbill/killbill-client-ruby/blob/master/spec/killbill_client/resource_spec.rb#L20) instead.

**Comment by pierre** [comment]
It would be good to run the [Ruby integration tests](https://github.com/killbill/killbill-integration-tests) (at least the `core` suite), to make sure this doesn't introduce any regression.

**Comment by daliwei** [comment]
@pierre Please review.

File: spec/killbill_client/encoding_spec.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
+require 'spec_helper'
+
+describe KillBillClient::API do
+  let(:expected_query_params) {
```
Comment (pierre) [inline spec/killbill_client/encoding_spec.rb]: This is still not deterministic: prior to 1.9, behavior of enumerated hashes was not in the ruby specification and therefore was up to implementation (in 1.9+, hash enumeration is specified by the language to be in the order of insertion). On Ruby 1.8, the output of `encode_params` is not predictable.

How about decoding `query_string` below?

```
url = 'http://www.foo.com?pluginProperty=details%3D%257B%2522eventType%2522%253A%2522voidEvent%2522%252C%2522transactionType%2522%253A%2522void%2522%257D'
puts CGI.parse(URI.parse(url).query)
{"pluginProperty"=>["details=%7B%22eventType%22%3A%22voidEvent%22%2C%22transactionType%22%3A%22void%22%7D"]}
```

Comment (daliwei) [inline spec/killbill_client/encoding_spec.rb]: @pierre Updated to decoding based spec. Please take a look.

**Comment by pierre** [comment]
Thanks!

**Pull request merged**